regression/cbmc/gcc_builtin_mul_overflow/main.c

```diff
@@ -0,0 +1,116 @@
+#include <assert.h>
+#include <limits.h>
+#include <stdint.h>
+
+#define A_VALUE_X_SO_THAT_X_TIMES_X_DOES_NOT_OVERFLOW(T)                       \
+  (((T)(1)) << (sizeof(T) * CHAR_BIT) / 2 - 1)
```

Comment:
Is this correct for both signed and unsigned?

Reply:
Left shift for positive signed is defined, as is left shift for unsigned. Either way, this results in `(T)2^(n/2-1)`, and `2^(n/2-1)*2^(n/2-1) == 2^(n-2)` (`n` being the width of the type) which fits comfortably into both signed and unsigned (assuming twos complement which we are in CBMC anyway).